cache/s3cache/s3cache.go

```diff
@@ -0,0 +1,242 @@
+	}
+	if maxZoom != 0 {
+		mz := uint(maxZoom)
+		s3cache.MaxZoom = &mz
+	}
+
+	s3cache.Bucket, err = c.String(ConfigKeyBucket, nil)
+	if err != nil {
+		return nil, ErrMissingBucket
+	}
+	if s3cache.Bucket == "" {
+		return nil, ErrMissingBucket
+	}
+
+	//	check for region env var
+	region := os.Getenv("AWS_REGION")
+	if region == "" {
+		region = DefaultRegion
+	}
+	region, err = c.String(ConfigKeyRegion, &region)
+	if err != nil {
+		return nil, err
+	}
+
+	accessKey := ""
+	accessKey, err = c.String(ConfigKeyAWSAccessKeyID, &accessKey)
+	if err != nil {
+		return nil, err
+	}
+	secretKey := ""
+	secretKey, err = c.String(ConfigKeyAWSSecretKey, &secretKey)
+	if err != nil {
+		return nil, err
+	}
+
+	awsConfig := aws.Config{
+		Region: aws.String(region),
+	}
+
+	//	support for static credentials, this is not recommended by AWS but
+	//	necessary for some environments
+	if accessKey != "" && secretKey != "" {
+		awsConfig.Credentials = credentials.NewStaticCredentials(accessKey, secretKey, "")
+	}
+
+	//	setup the s3 session.
+	//	if the accessKey and secreteKey are not provided (static creds) then the provider chain is used
+	//	http://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html
+	s3cache.Client = s3.New(
+		session.New(&awsConfig),
+	)
+
+	//	in order to confirm we have the correct permissions on the bucket create a small file
+	//	and test a PUT, GET and DELETE to the bucket
+	key := cache.Key{
+		MapName:   "tegola-test-map",
+		LayerName: "test-layer",
+		Z:         0,
+		X:         0,
+		Y:         0,
+	}
+	//	write a test file
+	if err := s3cache.Set(&key, []byte("\x53\x69\x6c\x61\x73")); err != nil {
+		return nil, err
+	}
+
+	//	read the test file
+	_, hit, err := s3cache.Get(&key)
+	if err != nil {
+		return nil, err
+	}
+	if !hit {
+		//	return an error?
+	}
+
+	//	purge the test file
+	if err := s3cache.Purge(&key); err != nil {
+		return nil, err
+	}
+
+	return &s3cache, nil
+}
+
+type S3Cache struct {
+	//	Bucket is the name of the s3 bucket to operate on
+	Bucket string
+
+	//	MaxZoom determins the max zoom the cache to persist. Beyond this
+	//	zoom, cache Set() calls will be ignored. This is useful if the cache
+	//	should not be leveraged for higher zooms when data changes often.
+	MaxZoom *uint
+
+	//	client holds a reference to the s3 client. it's expected the client
+	//	has an active session and read, write, delete permissions have been checked
+	Client *s3.S3
+}
+
+func (s3c *S3Cache) Set(key *cache.Key, val []byte) error {
+	var err error
+
+	//	check for maxzoom
+	if s3c.MaxZoom != nil && key.Z <= int(*s3c.MaxZoom) {
+		return nil
+	}
+
+	input := s3.PutObjectInput{
+		Body:   aws.ReadSeekCloser(bytes.NewReader(val)),
+		Bucket: aws.String(s3c.Bucket),
+		Key:    aws.String(key.String()),
+	}
+
+	_, err = s3c.Client.PutObject(&input)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			default:
+				return aerr
+			}
+		}
+		return err
+	}
+
+	return nil
```

Comment:
What is happening here is hidden. I think it would be easier to understand if it were:
```
return err
```

You are not doing anything special with the error here. The extra complication for the error transform and error handling is wasted.

Reply:
good point. I was leaving it stubbed out in case we discover an error to capture, but there is no point in being prepared for something that will likely never happen.